### PR TITLE
Easier access to state valuations, choice labels and choice origins

### LINKED ISSUE #178: How to access state and action labels from scheduler
Hi, I'm facing some problems when building a mdp model. I want to extract a scheduler from that but I'm not able to do it in the right way:
With storm I put this command and all works fine

```BASH
storm --prism mdp.pm --prop 'Pmax=? [F "done"]' --minmax:method vi --exportscheduler test.sched --buildstateval --buildchoiceorig --buildfull
```

with output:
```
___________________________________________________________________
Fully defined memoryless deterministic scheduler:
               model state:    choice(s)
    0: [p1=0	& p2=0	& p3=0]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 0)) & (p3 = 0)) -> 33/100 : (p1' = 1) + 33/100 : (p2' = 1) + 33/100 : (p3' = 1);)
    1: [p1=1	& p2=0	& p3=0]    BBB (robot: [BBB] (((p1 = 1) & (p2 = 0)) & (p3 = 0)) -> 1/2 : (p2' = 1) + 1/2 : (p3' = 1);)
    2: [p1=0	& p2=1	& p3=0]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 1)) & (p3 = 0)) -> 1/2 : (p1' = 1) + 1/2 : (p3' = 1);)
    3: [p1=0	& p2=0	& p3=1]    BBI (robot: [BBI] (((p1 = 0) & (p2 = 0)) & (p3 = 1)) -> 9/20 : (p1' = 1) + 9/20 : (p2' = 1) + 1/10 : (p3' = 2);)
    4: [p1=1	& p2=1	& p3=0]    BBB (robot: [BBB] (((p1 = 1) & (p2 = 1)) & (p3 = 0)) -> 1 : (p3' = 1);)
    5: [p1=1	& p2=0	& p3=1]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 0)) & (p3 = 1)) -> 1/20 : (p1' = 2) + 1/20 : (p3' = 2) + 9/10 : (p2' = 1);)
    6: [p1=2	& p2=0	& p3=0]    BBB (robot: [BBB] (((p1 = 2) & (p2 = 0)) & (p3 = 0)) -> 1/2 : (p2' = 1) + 1/2 : (p3' = 1);)
    7: [p1=0	& p2=1	& p3=1]    BBI (robot: [BBI] (((p1 = 0) & (p2 = 1)) & (p3 = 1)) -> 1/20 : (p2' = 2) + 1/20 : (p3' = 2) + 9/10 : (p1' = 1);)
    8: [p1=0	& p2=2	& p3=0]    BBT (robot: [BBT] (((p1 = 0) & (p2 = 2)) & (p3 = 0)) -> 9/20 : (p1' = 1) + 9/20 : (p3' = 1) + 1/10 : (p2' = 3);)
    9: [p1=0	& p2=0	& p3=2]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 0)) & (p3 = 2)) -> 1/2 : (p1' = 1) + 1/2 : (p2' = 1);)
   10: [p1=1	& p2=1	& p3=1]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 1)) & (p3 = 1)) -> 33/100 : (p1' = 2) + 33/100 : (p2' = 2) + 33/100 : (p3' = 2);)
   11: [p1=2	& p2=1	& p3=0]    BBB (robot: [BBB] (((p1 = 2) & (p2 = 1)) & (p3 = 0)) -> 1 : (p3' = 1);)
   12: [p1=1	& p2=2	& p3=0]    BBB (robot: [BBB] (((p1 = 1) & (p2 = 2)) & (p3 = 0)) -> 1 : (p3' = 1);)
   13: [p1=2	& p2=0	& p3=1]    BBI (robot: [BBI] (((p1 = 2) & (p2 = 0)) & (p3 = 1)) -> 9/10 : (p2' = 1) + 1/10 : (p3' = 2);)
   14: [p1=1	& p2=0	& p3=2]    BBB (robot: [BBB] (((p1 = 1) & (p2 = 0)) & (p3 = 2)) -> 1 : (p2' = 1);)
   15: [p1=3	& p2=0	& p3=0]    BBB (robot: [BBB] (((p1 = 3) & (p2 = 0)) & (p3 = 0)) -> 1/2 : (p2' = 1) + 1/2 : (p3' = 1);)
   16: [p1=0	& p2=2	& p3=1]    BBI (robot: [BBI] (((p1 = 0) & (p2 = 2)) & (p3 = 1)) -> 9/10 : (p1' = 1) + 1/10 : (p3' = 2);)
   17: [p1=0	& p2=1	& p3=2]    BBI (robot: [BBI] (((p1 = 0) & (p2 = 1)) & (p3 = 2)) -> 9/10 : (p1' = 1) + 1/10 : (p2' = 2);)
   18: [p1=0	& p2=3	& p3=0]    BIT (robot: [BIT] (((p1 = 0) & (p2 = 3)) & (p3 = 0)) -> 1/2 : (p1' = 1) + 1/2 : (p2' = 1);)
   19: [p1=0	& p2=0	& p3=3]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 0)) & (p3 = 3)) -> 1/2 : (p1' = 1) + 1/2 : (p2' = 1);)
   20: [p1=2	& p2=1	& p3=1]    BBI (robot: [BBI] (((p1 = 2) & (p2 = 1)) & (p3 = 1)) -> 1/2 : (p2' = 2) + 1/2 : (p3' = 2);)
   21: [p1=1	& p2=2	& p3=1]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 2)) & (p3 = 1)) -> 1/2 : (p1' = 2) + 1/2 : (p3' = 2);)
   22: [p1=1	& p2=1	& p3=2]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 1)) & (p3 = 2)) -> 1/2 : (p1' = 2) + 1/2 : (p2' = 2);)
   23: [p1=2	& p2=2	& p3=0]    BBB (robot: [BBB] (((p1 = 2) & (p2 = 2)) & (p3 = 0)) -> 1 : (p3' = 1);)
   24: [p1=3	& p2=1	& p3=0]    BBB (robot: [BBB] (((p1 = 3) & (p2 = 1)) & (p3 = 0)) -> 1 : (p3' = 1);)
   25: [p1=1	& p2=3	& p3=0]    BBB (robot: [BBB] (((p1 = 1) & (p2 = 3)) & (p3 = 0)) -> 1 : (p3' = 1);)
   26: [p1=2	& p2=0	& p3=2]    BBB (robot: [BBB] (((p1 = 2) & (p2 = 0)) & (p3 = 2)) -> 1 : (p2' = 1);)
   27: [p1=3	& p2=0	& p3=1]    BBI (robot: [BBI] (((p1 = 3) & (p2 = 0)) & (p3 = 1)) -> 9/10 : (p2' = 1) + 1/10 : (p3' = 2);)
   28: [p1=1	& p2=0	& p3=3]    BBB (robot: [BBB] (((p1 = 1) & (p2 = 0)) & (p3 = 3)) -> 1 : (p2' = 1);)
   29: [p1=3	& p2=3	& p3=0]    BBB (robot: [BBB] (((p1 = 3) & (p2 = 3)) & (p3 = 0)) -> 1 : (p3' = 1);)
   30: [p1=0	& p2=2	& p3=2]    BBT (robot: [BBT] (((p1 = 0) & (p2 = 2)) & (p3 = 2)) -> 1/20 : (p2' = 3) + 1/20 : (p3' = 3) + 9/10 : (p1' = 1);)
   31: [p1=0	& p2=3	& p3=1]    BBI (robot: [BBI] (((p1 = 0) & (p2 = 3)) & (p3 = 1)) -> 9/10 : (p1' = 1) + 1/10 : (p3' = 2);)
   32: [p1=0	& p2=1	& p3=3]    BBI (robot: [BBI] (((p1 = 0) & (p2 = 1)) & (p3 = 3)) -> 9/10 : (p1' = 1) + 1/10 : (p2' = 2);)
   33: [p1=2	& p2=2	& p3=1]    BBI (robot: [BBI] (((p1 = 2) & (p2 = 2)) & (p3 = 1)) -> 1 : (p3' = 2);)
   34: [p1=2	& p2=1	& p3=2]    BBI (robot: [BBI] (((p1 = 2) & (p2 = 1)) & (p3 = 2)) -> 1 : (p2' = 2);)
   35: [p1=3	& p2=1	& p3=1]    BBI (robot: [BBI] (((p1 = 3) & (p2 = 1)) & (p3 = 1)) -> 1/2 : (p2' = 2) + 1/2 : (p3' = 2);)
   36: [p1=1	& p2=2	& p3=2]    BBT (robot: [BBT] (((p1 = 1) & (p2 = 2)) & (p3 = 2)) -> 1/2 : (p2' = 3) + 1/2 : (p3' = 3);)
   37: [p1=1	& p2=3	& p3=1]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 3)) & (p3 = 1)) -> 1/2 : (p1' = 2) + 1/2 : (p3' = 2);)
   38: [p1=1	& p2=1	& p3=3]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 1)) & (p3 = 3)) -> 1/2 : (p1' = 2) + 1/2 : (p2' = 2);)
   39: [p1=3	& p2=2	& p3=0]    BBB (robot: [BBB] (((p1 = 3) & (p2 = 2)) & (p3 = 0)) -> 1 : (p3' = 1);)
   40: [p1=2	& p2=3	& p3=0]    BBB (robot: [BBB] (((p1 = 2) & (p2 = 3)) & (p3 = 0)) -> 1 : (p3' = 1);)
   41: [p1=3	& p2=0	& p3=2]    BBB (robot: [BBB] (((p1 = 3) & (p2 = 0)) & (p3 = 2)) -> 1 : (p2' = 1);)
   42: [p1=2	& p2=0	& p3=3]    BBB (robot: [BBB] (((p1 = 2) & (p2 = 0)) & (p3 = 3)) -> 1 : (p2' = 1);)
   43: [p1=3	& p2=3	& p3=1]    BBI (robot: [BBI] (((p1 = 3) & (p2 = 3)) & (p3 = 1)) -> 1 : (p3' = 2);)
   44: [p1=0	& p2=3	& p3=2]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 3)) & (p3 = 2)) -> 1 : (p1' = 1);)
   45: [p1=0	& p2=2	& p3=3]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 2)) & (p3 = 3)) -> 1 : (p1' = 1);)
   46: [p1=2	& p2=2	& p3=2]    BBT (robot: [BBT] (((p1 = 2) & (p2 = 2)) & (p3 = 2)) -> 33/100 : (p1' = 3) + 33/100 : (p2' = 3) + 33/100 : (p3' = 3);)
   47: [p1=3	& p2=2	& p3=1]    BBI (robot: [BBI] (((p1 = 3) & (p2 = 2)) & (p3 = 1)) -> 1 : (p3' = 2);)
   48: [p1=2	& p2=3	& p3=1]    BBI (robot: [BBI] (((p1 = 2) & (p2 = 3)) & (p3 = 1)) -> 1 : (p3' = 2);)
   49: [p1=3	& p2=1	& p3=2]    BBI (robot: [BBI] (((p1 = 3) & (p2 = 1)) & (p3 = 2)) -> 1 : (p2' = 2);)
   50: [p1=2	& p2=1	& p3=3]    BBI (robot: [BBI] (((p1 = 2) & (p2 = 1)) & (p3 = 3)) -> 1 : (p2' = 2);)
   51: [p1=1	& p2=3	& p3=2]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 3)) & (p3 = 2)) -> 1 : (p1' = 2);)
   52: [p1=1	& p2=2	& p3=3]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 2)) & (p3 = 3)) -> 1 : (p1' = 2);)
   53: [p1=3	& p2=0	& p3=3]    BBB (robot: [BBB] (((p1 = 3) & (p2 = 0)) & (p3 = 3)) -> 1 : (p2' = 1);)
   54: [p1=3	& p2=3	& p3=2]    BBT (robot: [BBT] (((p1 = 3) & (p2 = 3)) & (p3 = 2)) -> 1 : (p3' = 3);)
   55: [p1=0	& p2=3	& p3=3]    BBB (robot: [BBB] (((p1 = 0) & (p2 = 3)) & (p3 = 3)) -> 1 : (p1' = 1);)
   56: [p1=3	& p2=2	& p3=2]    BBT (robot: [BBT] (((p1 = 3) & (p2 = 2)) & (p3 = 2)) -> 1/2 : (p2' = 3) + 1/2 : (p3' = 3);)
   57: [p1=2	& p2=3	& p3=2]    BBT (robot: [BBT] (((p1 = 2) & (p2 = 3)) & (p3 = 2)) -> 1/2 : (p1' = 3) + 1/2 : (p3' = 3);)
   58: [p1=2	& p2=2	& p3=3]    BBT (robot: [BBT] (((p1 = 2) & (p2 = 2)) & (p3 = 3)) -> 1/2 : (p1' = 3) + 1/2 : (p2' = 3);)
   59: [p1=3	& p2=1	& p3=3]    BBI (robot: [BBI] (((p1 = 3) & (p2 = 1)) & (p3 = 3)) -> 1 : (p2' = 2);)
   60: [p1=1	& p2=3	& p3=3]    BBI (robot: [BBI] (((p1 = 1) & (p2 = 3)) & (p3 = 3)) -> 1 : (p1' = 2);)
   61: [p1=3	& p2=3	& p3=3]    No origin
   62: [p1=3	& p2=2	& p3=3]    BBT (robot: [BBT] (((p1 = 3) & (p2 = 2)) & (p3 = 3)) -> 1 : (p2' = 3);)
   63: [p1=2	& p2=3	& p3=3]    BBT (robot: [BBT] (((p1 = 2) & (p2 = 3)) & (p3 = 3)) -> 1 : (p1' = 3);)
___________________________________________________________________

```
How I can do this in stormpy? I fo this:

```PYTHON
prism_program = stormpy.parse_prism_program(path)
    formula = "Pmax=? [ F \"done\" ]"
    f = stormpy.parse_properties(formula, prism_program)
    
    # options = stormpy.ModelCheckerOptions()
    options = stormpy.BuilderOptions([p.raw_formula for p in f])
    options.set_build_state_valuations(True)
    options.set_build_with_choice_origins(True)
    options.set_build_choice_labels()
    options.set_build_all_labels()
    
    model = stormpy.build_sparse_model_with_options(prism_program, options)
    result = stormpy.model_checking(model, f[0], extract_scheduler=True)
    assert result.has_scheduler

    scheduler = result.scheduler
    assert scheduler.memoryless
    assert scheduler.deterministic
    print(scheduler)
```

with output:

```
___________________________________________________________________
Fully defined memoryless deterministic scheduler:
model state:    choice(s)
           0    0
           1    0
           2    0
           3    1
           4    0
           5    1
           6    0
           7    1
           8    2
           9    0
          10    0
          11    0
          12    0
          13    1
          14    0
          15    0
          16    1
          17    1
          18    5
          19    0
          20    0
          21    0
          22    0
          23    0
          24    0
          25    0
          26    0
          27    1
          28    0
          29    0
          30    2
          31    1
          32    1
          33    0
          34    0
          35    0
          36    1
          37    0
          38    0
          39    0
          40    0
          41    0
          42    0
          43    0
          44    0
          45    0
          46    0
          47    0
          48    0
          49    0
          50    0
          51    0
          52    0
          53    0
          54    0
          55    0
          56    0
          57    0
          58    0
          59    0
          60    0
          61    0
          62    0
          63    0
___________________________________________________________________
```

It do not show the labels on the state and actions. How can I solve it? Thanks

### PR DESCRIPTION
**Pull request by volkm**
Allows to access labels, valuations and origins while iterating over states and actions.

Resolves #178

### CODE REVIEW

File: src/storage/state.cpp [hunk 1]
```diff
@@ -13,7 +13,8 @@ void define_state(py::module& m, std::string const& vtSuffix) {
     py::class_<SparseModelState<ValueType>>(m, ("Sparse" + vtSuffix + "ModelState").c_str(), "State in sparse model")
         .def("__str__", &SparseModelState<ValueType>::toString)
         .def_property_readonly("id", &SparseModelState<ValueType>::getIndex, "Id")
-        .def_property_readonly("labels", &SparseModelState<ValueType>::getLabels, "Labels")
+        .def_property_readonly("labels", &SparseModelState<ValueType>::getLabels, "Get state labels")
+        .def_property_readonly("valuations", &SparseModelState<ValueType>::getValuations, "Get state valuations")
```
Comment (sjunges) [inline src/storage/state.cpp:17]: Wasnt there something that this is superslow as it creates a copy?

Comment (volkm) [reply src/storage/state.cpp:17]: Can you elaborate what exactly makes it superslow? The `readonly` or the general implementation of the functions `getLabels`?
In general, the iteration over states / actions can certainly be improved peformance-wise. One issue for instance is that we have to keep track of the model. Maybe we open a separate issue for the performance and discuss options there?

Comment (sjunges) [reply src/storage/state.cpp:17]: Ah, I misread it initially as if it was generating a copy of the complete state valuations everytime. That does not seem to happen. So indeed, we can merge it. We may want to discuss elsewhere whether we want to move the iteration over states with this slower high-level API to the stormvogel api.

**Comment by sjunges** [comment]
LGTM. Sorry for the confusion, many thanks for the contribution.

**Pull request merged**